lib/spack/spack/util/compression.py

```diff
@@ -285,33 +309,346 @@ def decompressor_for(path, ext):
+    a magic number or extension string.
+
+    Subclasses should each describe a different
+    type of file. In order to do so, they must define
+    the extension string, magic number, and header offset (if non zero).
+    If a class has multiple magic numbers, it will need to
+    override the method describin that file types magic numbers and
+    the method that checks a types magic numbers against a given file's.
+    """
+
+    _OFFSET = 0
+    _compressed = False
+
+    @classmethod
+    def name(cls):
+        raise NotImplementedError
+
+    @classmethod
+    def is_compression_type(cls):
+        return cls._compressed
+
+    @classmethod
+    def offset(cls):
+        return cls._OFFSET
+
+    @classmethod
+    def magic_number(cls):
+        return cls._MAGIC_NUMBER
+
+    @classmethod
+    def extension(cls):
+        return cls._ext
+
+    @classmethod
+    def header_size(cls):
+        return len(cls.magic_number())
+
+    @classmethod
+    def _bytes_check(cls, magic_bytes):
+        return cls._MAGIC_NUMBER == magic_bytes
+
+    @classmethod
+    def is_file_of_type(cls, iostream):
+        """Query byte stream for appropriate magic number
+
+        Args:
+            iostream: file byte stream
+
+        Returns:
+            Bool denoting whether file is of class file type
+            based on magic number
+        """
+        if not iostream:
+            return False
+        # move to location of magic bytes
+        iostream.seek(cls.offset())
+        magic_bytes = iostream.read(cls.header_size())
+        # return to beginning of file
+        iostream.seek(0)
+        if cls._bytes_check(magic_bytes):
+            return True
+        return False
+
+
+class CompressedFileTypeInterface(FileTypeInterface):
+    """Interface class for FileTypes that include compression information"""
+
+    _compressed = True
+
+    @classmethod
+    def decomp_in_memory(cls, stream):
+        raise NotImplementedError("Implementation by compression subclass required")
+
+
+class BZipFileType(CompressedFileTypeInterface):
+    _MAGIC_NUMBER = b"\x42\x5a\x68"
+    _ext = "bz2"
+
+    @classmethod
+    def name(cls):
+        return "bzip2 compressed data"
+
+    @classmethod
+    def decomp_in_memory(cls, stream):
+        if is_bz2_supported():
+            comp_stream = stream.read(TarFileType.offset() + TarFileType.header_size())
+            return io.BytesIO(initial_bytes=bz2.BZ2Decompressor().decompress(comp_stream))
+        return None
+
+
+class ZCompressedFileType(CompressedFileTypeInterface):
+    _MAGIC_NUMBER_LZW = b"\x1f\x9d"
+    _MAGIC_NUMBER_LZH = b"\x1f\xa0"
+    _ext = "Z"
+
+    @classmethod
+    def name(cls):
+        return "compress'd data"
+
+    @classmethod
+    def magic_number(cls):
+        return [cls._MAGIC_NUMBER_LZH, cls._MAGIC_NUMBER_LZW]
+
+    @classmethod
+    def header_size(cls):
+        return max(len(cls._MAGIC_NUMBER_LZW), len(cls._MAGIC_NUMBER_LZH))
```

Comment:
You are prepped for these being different lengths, but if they are, one of them will be longer, in which case I assume we'd pull enough bytes for the longer one (which is a problem because we don't do prefix matching and just check if `==`).

Reply:
Addressed.